// next-intl/dist/src/server/react-client/index.js
function notSupported(name) {
    throw new Error(`\`${name}\` is not supported in Client Components.`);
}
// Must match `../index.tsx`
// prettier-ignore
export const getRequestConfig = notSupported('getRequestConfig');
// prettier-ignore
export const getIntl = notSupported('getIntl');
// prettier-ignore
export const getFormatter = notSupported('getFormatter');
// prettier-ignore
export const getLocale = notSupported('getLocale');
// prettier-ignore
export const getNow = notSupported('getNow');
// prettier-ignore
export const getTimeZone = notSupported('getTimeZone');
// prettier-ignore
export const getTranslations = notSupported('getTranslations');
// prettier-ignore
export const getTranslator = notSupported('getTranslator');
// prettier-ignore
export const getMessages = notSupported('getMessages');
// TODO: Since this is available in Client Comonents too, we should really
// consider exporting this from `next-intl/navigation` instead. For now, for
// compatibility, we'll add a client entry point for `next-intl/server`.
export { default as redirect } from './redirect';
